pkg/resource/deploy/step_generator.go

```diff
@@ -399,75 +400,153 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, *re
+	condemnedResources := make(graph.ResourceSet)
+	for i := len(prev.Resources) - 1; i >= 0; i-- {
+		res := prev.Resources[i]
+		if res.Delete {
+			logging.V(7).Infof(
+				"stepGenerator.GeneratePendingDeletes(): resource (%v, %v) is pending deletion", res.URN, res.ID)
+			sg.pendingDeletes[res] = true
+			condemnedResources[res] = true
+		}
+	}
+
+	return sg.scheduleDeletes(condemnedResources, true /* pendingDeletesAreReplaces */)
+}
+
+// scheduleDeletes takes a set of resources that are condemned for deletion and "schedules" them by producing a list of
+// list of steps, where each list can be executed in parallel but a previous list must be executed to completion before
+// advacing to the next list.
+//
+// In lieu of tracking per-step dependencies and orienting the step executor around these dependencies, this function
+// provides a conservative approximation of what deletions can safely occur in parallel. The insight here is that the
+// resource dependency graph is a partially-ordered set and all partially-ordered sets can be easily decomposed into
+// antichains - subsets of the set that are all not comparable to one another. (In this definition, "not comparable"
+// means "do not depend on one another").
+//
+// The algorithm for decomposing a poset into antichains is:
+//  1. While there exist elements in the poset,
+//    1a. There must exist at least one "maximal" element of the poset. Let E_max be those elements.
+//    2a. Remove all elements E_max from the poset. E_max is an antichain.
+//    3a. Goto 1.
+//
+// Translated to our dependency graph:
+//  1. While the set of condemned resources is not empty:
+//    1a. Remove all resources with no outgoing edges from the graph and add them to the current antichain.
+//    2a. Goto 1.
+//
+// The resulting list of antichains is a list of list of steps that can be safely executed in parallel. Since we must
+// process deletes in reverse (so we don't delete resources upon which other resources depend), we reverse the list and
+// hand it back to the plan executor for safe execution.
+func (sg *stepGenerator) scheduleDeletes(condemned graph.ResourceSet, pendingDeletesAreReplaces bool) []Antichain {
+	var antichains []Antichain
+	dg := sg.plan.depGraph
+	for len(condemned) > 0 {
+		var antichain Antichain
+		logging.V(7).Infof("Planner beginning schedule of new deletion antichain")
+		for res := range condemned {
+			// Does res have any outgoing edges to resources that haven't already been removed from the graph?
+			// For the purposes of deletion calculation, res.Parent counts as an outgoing edge.
+			condemnedDependencies := dg.DependenciesOf(res).Intersect(condemned)
+			if len(condemnedDependencies) == 0 && !condemned[dg.ParentOf(res)] {
+				// If not, it's safe to delete res at this stage.
+				logging.V(7).Infof("Planner scheduling deletion of '%v'", res.URN)
+				if res.Delete && !pendingDeletesAreReplaces {
```

Comment:
I accidentally misnamed `pendingDeletesAreReplaces` here (I renamed it just before opening the PR). The polarity of this bool should be reversed.

Reply:
Will fix in the next commit.